File: algorithms_and_data_structures/Java_Versions/Barbosa_P3/Deck.java
import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

public class Deck
{
    // default constructor
	public Deck()
    {
        numCards = 52;
        theDeck  = new Card[numCards];
        for(int i = 0; i < 13; ++i)
            theDeck[i] = new Card((i + 1)%13, "clubs");
        for(int i = 13; i < 26; ++i)
            theDeck[i] = new Card((i + 1)%13, "spades");
        for(int i = 26; i < 39; ++i)
            theDeck[i] = new Card((i + 1)%13, "hearts");
        for(int i = 39; i < numCards; ++i)
            theDeck[i] = new Card((i + 1)%13, "diamonds");
        topCard  = 52;
    }

    // Remove the top card from the deck and return it.
	public Card dealCard()
    {
        return theDeck[topCard--];
    }

    // Shuffle the cards in the deck
	public void Shuffle()
    {
        Random rnd = ThreadLocalRandom.current();
        for(int i = numCards - 1; i > 0; --i)
        {
            int index      = rnd.nextInt(i + 1);
            Card c         = theDeck[index];
            theDeck[index] = theDeck[i];
            theDeck[i]     = c;

        }
    }

    // return true if there are no more cards in the deck, false otherwise
	public boolean isEmpty()
    {
        return topCard < 0;
    }

    //overload << operator to display the deck
    public void Print()
    {
        for(Card c : theDeck)
            c.Print();
    }

    public static void main(String[] args)
    {
        Deck d = new Deck();
        d.Print();
        d.Shuffle();
        d.Print();
    }

    private int    numCards;
	private Card[] theDeck;
	private int    topCard;					// the index of the deck's top card
};

File: algorithms_and_data_structures/Project3/Deck.cpp
////Author:      Ryan Florida
////Project:     OLA3
////Due Date:    February 22, 2017
////Description: This is the implementation file for the Deck class.

#include "Deck.h"
#include <ctime>
#include <cstdlib>
#include <iomanip>

using std::cout;
using std::setw;
using std::left;

//Constructor.
Deck::Deck()
    :topCard(0)
{
    //Initialize entire deck.
    for(int i = 0; i < 13; ++i)
    {
        theDeck[i]      = *(new Card(i + 2, clubs));
        theDeck[i + 13] = *(new Card(i + 2, hearts));
        theDeck[i + 26] = *(new Card(i + 2, diamonds));
        theDeck[i + 39] = *(new Card(i + 2, spades));
    }
}

//Remove the top card from the deck and return it.
Card Deck::dealCard()
{
    int currentTop = topCard++;

    return theDeck[currentTop];
}

//Shuffle the cards in the deck.
void Deck::Shuffle()
{
    //Output the default deck configuration.
    cout << "The original deck has the following cards: \n";
    for(int i = 0; i < Card_Num; i++)
    {
        if(i != 0 && i%13 == 0)
            cout << '\n';

        switch(theDeck[i].getFaceValue())
        {
            case 11:
                cout << 'J';
                break;
            case 12:
                cout << 'Q';
                break;
            case 13:
                cout << 'K';
                break;
            case 14:
                cout << 'A';
                break;
            default:
                cout << theDeck[i].getFaceValue();
        }

        theDeck[i].displaySuite();
        cout << '[' << theDeck[i].getPointValue() << ']' << left << setw(3)
             << ' ';
    }

    //Shuffle based on random seed value.
    srand((unsigned)time(0));

    Card temp;
    int randomIndex1,
        randomIndex2,
        randomNumber = rand()%10000 + 1000000;

    //Perform a healthy shuffle.
    for(int i = 0; i < randomNumber; i++)
    {
        randomIndex1 = rand()%52;
        do
        {
            randomIndex2 = rand()%52;
        }while(randomIndex2 == randomIndex1);

        temp                  = theDeck[randomIndex1];
        theDeck[randomIndex1] = theDeck[randomIndex2];
        theDeck[randomIndex2] = temp;
    }

    //Display the newly shuffled deck.
    cout << "\n\nThe shuffled deck has the following cards: \n";
    for(int i = 0; i < Card_Num; i++)
    {
        if(i != 0 && i%13 == 0)
            cout << '\n';

        switch(theDeck[i].getFaceValue())
        {
            case 11:
                cout << 'J';
                break;
            case 12:
                cout << 'Q';
                break;
            case 13:
                cout << 'K';
                break;
            case 14:
                cout << 'A';
                break;
            default:
                cout << theDeck[i].getFaceValue();
        }

        theDeck[i].displaySuite();
        cout << '[' << theDeck[i].getPointValue() << ']' << left << setw(3)
             << ' ';
    }

    cout << '\n';

    return;
}

//Return true if the deck has run out of cards. Return false otherwise.
bool Deck::isEmpty()
{
    return (topCard == Card_Num);
}
